draft-ietf-quic-version-negotiation.md

```diff
@@ -67,105 +67,171 @@ Discussion of this work is encouraged to happen on the QUIC IETF mailing list
+and one "compatible" that allows saving the round trip but only applies when
+the versions are compatible.
+
+The client initiates a QUIC connection by sending a first flight of QUIC
+packets with a long header to the server {{INV}}. We'll refer to the version of
+those packets as the "original version". The client's first flight includes
+handshake version information (see {{hs-vers-info}}) which will be used to
+optionally enable compatible version negotation (see {{compat-vn}}), and to
+prevent version downgrade attacks (see {{downgrade}}).
+
+Upon receiving this first flight, the server verifies whether it knows how to
+parse first flights from the original version. If it does not, then it starts
+incompatible version negotiation, see {{incompat-vn}}. If the server can parse
+the first flight, it can either establish the connection using the original
+version, or it MAY attempt compatible version negotiation, see {{compat-vn}}.
+
+Note that it is possible for a server to have the ability to parse the first
+flight of a given version without fully supporting it, in the sense that it
+implements enough of the version's specification to parse first flight packets
+but not enough to fully establish a connection using that version.
+
+
+## Incompatible Version Negotiation {#incompat-vn}
+
+The server starts incompatible version negotiation by sending a VN packet,
+listing all the versions that it does support.
+
+Upon receiving the VN packet, the client will search for a version it supports
+in the list provided by the server. If it doesn't find one, it aborts the
+connection attempt. Otherwise, it selects a mutually supported version and
+sends a new first flight with that version - we refer to this version as the
+"negotiated version".
+
+When the client sends its new first flight, it SHOULD use the same 5-tuple
+(source and destination IP addresses and UDP port numbers) and the same
+Destination Connection ID as its original first flight. This will increase the
+likelihood of both first flights reaching the same server.
+
+The new first flight will allow the endpoints to establish a connection using
+the negotiated version. The handshake of the negotiated version will exchange
+handshake version information (see {{hs-vers-info}}) required to ensure that VN
+was genuine, i.e. that no attacker injected packets in order to influence the
+VN process, see {{downgrade}}.
+
+
+## Compatible Version Negotiation {#compat-vn}
+
+When the server can parse the client's first flight using the original version,
+it can extract the client's handshake version information (see
+{{hs-vers-info}}). This contains the list of versions that the client thinks
+its original version is compatible with.
+
+If the server supports one of the client's compatible versions, and the server
+also believes that the original version is compatible with this version, then
+the server attempts to convert the client's first flight to that version. If
```

Comment:
As above, I think this whole "convert" thing is problematic. We should confine ourselves to Initials which are valid for both versions.

Reply:
What does valid mean? In my understanding of the word, an initial from version A will never be valid in version B because version B requires the version on the long header to be B.

Reply:
@ekr said:
> Not necessarily. We could say that for version B, the long header could have either A or B. This basically requires reinterpreting LongHeader.Version as "format" or "minimum version to interpret" or something, but it's not inherently a problem.

I think we should decouple documents from compatibility. It should be possible to have versions A and B not know about each other, and then someone writes the compatibility doc between them later.

But there are other reasons for things to not be valid: take `h3-32` and `h3-33` for example - they use a different TLS extension codepoints (0xffa5 vs 57) for transport parameters. That makes a first flight from one invalid in the other. However, it's possible to convert from one to the other. That's why I think of conversion instead of validity.